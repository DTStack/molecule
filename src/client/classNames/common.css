*:focus-visible {
    outline: none;
}

body [tabindex='0']:focus,
body [tabindex='-1']:focus,
body .synthetic-focus,
body select:focus,
body input[type='button']:focus,
body input[type='text']:focus,
body button:focus,
body textarea:focus,
body input[type='search']:focus,
body input[type='checkbox']:focus {
    outline-width: 1px;
    outline-style: solid;
    outline-offset: -1px;
    outline-color: var(--focusBorder);
    opacity: 1;
}

body .synthetic-focus :focus {
    outline: 0 !important; /* elements within widgets that draw synthetic-focus should never show focus */
}

body .monaco-inputbox.info.synthetic-focus,
body .monaco-inputbox.warning.synthetic-focus,
body .monaco-inputbox.error.synthetic-focus,
body .monaco-inputbox.info input[type='text']:focus,
body .monaco-inputbox.warning input[type='text']:focus,
body .monaco-inputbox.error input[type='text']:focus {
    outline: 0 !important; /* outline is not going well with decoration */
}

body .monaco-list:focus {
    outline: 0 !important; /* tree indicates focus not via outline but through the focused item */
}

body {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Noto Sans', Helvetica, Arial,
        sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji';
}
