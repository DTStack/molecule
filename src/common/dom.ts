export type HTMLElementType = HTMLElement | null;
export type TriggerEvent = 'click' | 'contextmenu' | 'hover';
/**
 * specify `rightBottom` means align to the bottom and keep in right
 */
export type PlacementType = 'top' | 'right' | 'bottom' | 'left' | 'rightBottom';
export const select = document.querySelector.bind(document);
export const selectAll = document.querySelectorAll.bind(document);

export interface IPosition {
    x: number;
    y: number;
}

/**
 * Get Document Rectangle info
 */
export function getDocumentRect() {
    const body = document.body;
    const html = document.documentElement;

    const clientWidth = html.clientWidth;
    const clientHeight = html.clientHeight;

    const height = Math.max(
        body.scrollHeight,
        body.offsetHeight,
        clientHeight,
        html.scrollHeight,
        html.offsetHeight
    );

    const width = Math.max(
        body.scrollWidth,
        body.offsetWidth,
        clientWidth,
        html.scrollWidth,
        html.offsetWidth
    );

    return {
        height,
        width,
        clientWidth,
        clientHeight,
    };
}

/**
 * Returns the position of element relative to element position
 * @param element target element
 * @param relativePos the relative element position
 */
export function getRelativePosition(
    element: HTMLElement,
    relativePos: IPosition
) {
    const { clientWidth, clientHeight } = getDocumentRect();
    const anchorX = relativePos.x;
    const anchorY = relativePos.y;

    const marginRight = clientWidth - anchorX;
    const marginBottom = clientHeight - anchorY;

    const viewHeight = element.offsetHeight;
    const viewWidth = element.offsetWidth;

    const x = marginRight < viewWidth ? anchorX - viewWidth : anchorX;
    const y = marginBottom < viewHeight ? anchorY - viewHeight : anchorY;

    return {
        x,
        y,
    };
}

export function getEventPosition(e: React.MouseEvent) {
    return {
        x: e.clientX,
        y: e.clientY,
    };
}

export function findParentByClassName<T>(element, className): T | null {
    try {
        while (element && element) {
            const classes = element.getAttribute('class');
            if (classes && classes.indexOf(className) > -1) {
                return element;
            }
            element = element.parentElement;
        }
    } catch (e) {
        throw new Error(`findParentByClassName error: ${e}`);
    }
    return null;
}

export function triggerEvent(trigger: TriggerEvent) {
    switch (trigger) {
        case 'click': {
            return 'onClick';
        }
        case 'hover': {
            return 'onMouseOver';
        }
        case 'contextmenu': {
            return 'onContextMenu';
        }
        default: {
            return trigger;
        }
    }
}

/**
 * Get the element position by placement and DOMRect
 * @param placement top | right | bottom | left
 * @param domRect Dom rect info, normally get it from getBoundingClientRect function
 */
export function getPositionByPlacement(
    placement: PlacementType,
    domRect: DOMRect
): IPosition {
    let x = domRect.x; // Initial placement is top
    let y = domRect.y;

    if (placement === 'top') {
        y = domRect.y - domRect.height;
    } else if (placement === 'right') {
        x = domRect.x + domRect.width;
    } else if (placement === 'bottom') {
        y = domRect.y + domRect.height;
    } else if (placement === 'left') {
        x = domRect.x - domRect.width;
    } else if (placement === 'rightBottom') {
        x = domRect.x + domRect.width;
        y = domRect.y + domRect.height;
    }
    return { x, y };
}

export function getAttr(domElement: HTMLElement, attr) {
    return domElement.getAttribute(attr) || '';
}

/**
 * Get an element the center coords
 * @param element HTMLElement
 * @returns
 */
export function getElementClientCenter(element: HTMLElement) {
    const { left, top, width, height } = element.getBoundingClientRect();
    return {
        x: left + width / 2,
        y: top + height / 2,
    };
}

/**
 * Get the data-* attributions from props
 * @param props
 * @returns
 */
export function getDataAttributionsFromProps(
    props: Record<string, any>
): Record<string, any> {
    return Object.keys(props).reduce((pre, cur) => {
        if (cur.startsWith('data-')) {
            pre[cur] = props[cur];
        }
        return pre;
    }, {});
}
