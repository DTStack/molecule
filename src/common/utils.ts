export function searchById(id) {
    return (item) => item.id === id;
}

/**
 * Clone a new object by an object instance
 * @param origin Original object instance
 */
export function cloneInstance<T>(origin: T) {
    try {
        const prototypes = Object.getPrototypeOf(origin) || {};
        Object.keys(prototypes).forEach((prop) => {
            if (typeof prototypes[prop] === 'function') {
                prototypes[prop].bind(origin);
            }
        });
        return {
            ...origin,
            ...prototypes,
        };
    } catch (e) {
        console.error('Function cloneInstance error:', e);
    }
}

/**
 * Merge multiple functions to one function
 * @param funcs
 */
export function mergeFunctions(...funcs) {
    return function (...args) {
        funcs.filter((fn) => !!fn).forEach((fn) => fn?.(...args));
    };
}

export function randomId() {
    return Date.now() + Math.round(Math.random() * 1000);
}

export function mergeObjects<T>(source: object, target: object): T {
    for (const prop in target) {
        if (typeof source[prop] === 'object' && target[prop]) {
            source[prop] = mergeObjects(source[prop], target[prop]);
        } else {
            source[prop] = target[prop];
        }
    }
    return source as any;
}

/**
 * It's converts an object to a flatted object,
 * eg: { a: { b: 'test' }}, result is : { 'a.b': 'test' }
 * @param target flat target
 */
export function flatObject(target: object): object {
    const flatted = {};
    const recurse = (parent: string, target: object) => {
        for (const prop in target) {
            if (prop) {
                const value = target[prop];
                const path = parent ? `${parent}.${prop}` : prop;
                if (typeof value === 'object') {
                    recurse(path, value);
                } else {
                    flatted[path] = value;
                }
            }
        }
    };
    recurse('', target);
    return flatted;
}

/**
 * It's converts a flatted object to a normal object,
 *  eg: { 'a.b': 'test' }, result is : { a: { b: 'test' }}
 * @param target flat target
 */
export function normalizeFlattedObject(target: object): object {
    let normalized = {};

    for (const prop in target) {
        if (prop) {
            const flattedProps = prop.split('.');
            let prevProp = '';
            let root = {};
            let prevObj = root;
            for (let i = 0; i < flattedProps.length; ++i) {
                if (i === flattedProps.length) break;
                const key = flattedProps[i];
                const current = { [key]: '' };

                if (prevProp) {
                    prevObj[prevProp] = current;
                }
                prevObj = current;
                prevProp = key;
                if (i === 0) {
                    root = current;
                }
            }
            prevObj[prevProp] = target[prop];
            normalized = mergeObjects(normalized, root);
        }
    }
    return normalized;
}

/**
 * Determine if a color is light or dark.
 * @param color HEX or RGB
 */
export function colorLightOrDark(color: string) {
    // Variables for red, green, blue values
    let r: number;
    let g: number;
    let b: number;

    // Check the format of the color, HEX or RGB?
    if (color.match(/^rgb/)) {
        // If RGB --> store the red, green, blue values in separate variables
        const matchArray =
            color.match(
                /^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/
            ) || [];
        r = +matchArray[1];
        g = +matchArray[2];
        b = +matchArray[3];
    } else {
        // If hex --> Convert it to RGB
        let rgbNum = +('0x' + color.slice(1, 7));
        if (color.length < 5) {
            rgbNum = +('0x' + color.slice(1).replace(/./g, '$&$&').slice(0, 6));
        }
        r = rgbNum >> 16;
        g = (rgbNum >> 8) & 255;
        b = rgbNum & 255;
    }

    // HSP (Highly Sensitive Poo) equation from http://alienryderflex.com/hsp.html
    const hsp = Math.sqrt(0.299 * (r * r) + 0.587 * (g * g) + 0.114 * (b * b));

    // Using the HSP value, determine whether the color is light or dark
    if (hsp > 127.5) {
        return 'light';
    }
    return 'dark';
}
