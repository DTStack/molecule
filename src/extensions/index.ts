import { ExtendsActions } from './actions';
import { ExtendsActivityBar } from './activityBar';
import { ExtendsAuxiliaryBar } from './auxiliaryBar';
import { ExtendsContextMenu } from './contextMenu';
import { ExtendsEditor } from './editor';
import { ExtendsEditorTree } from './editorTree';
import { ExtendsExplorer } from './explorer';
import { ExtendsFolderTree } from './folderTree';
import { LocalesExtends } from './locales';
import { ExtendsMenuBar } from './menuBar';
import { ExtendsNotification } from './notification';
import { ExtendsPanel } from './panel';
import { ExtendsSearch } from './search';
import { ExtendsSettings } from './settings';
import { ExtendsSidebar } from './sidebar';
import { ExtendsStatusBar } from './statusBar';
import { ColorThemeExtension } from './themes';

export default [
    LocalesExtends,
    ColorThemeExtension,
    ExtendsActions,
    ExtendsEditor,
    ExtendsActivityBar,
    ExtendsEditorTree,
    ExtendsPanel,
    ExtendsMenuBar,
    ExtendsNotification,
    ExtendsStatusBar,
    ExtendsSearch,
    ExtendsSettings,
    ExtendsFolderTree,
    ExtendsSidebar,
    ExtendsExplorer,
    ExtendsContextMenu,
    ExtendsAuxiliaryBar,
];
