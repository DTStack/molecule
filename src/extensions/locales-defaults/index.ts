import { IExtension, IContributeType } from 'mo/model/extension';

const koKR = require('./locales/ko-KR.json');
const zhCN = require('./locales/zh-CN.json');
const en = require('./locales/en.json');
const locales = [zhCN, en, koKR];

export const ExtendsLocales: IExtension = {
    id: 'ExtendsLocales',
    name: 'Extends locales',
    contributes: {
        [IContributeType.Languages]: locales,
    },
    activate() {},
    dispose() {},
};
export const BuiltInLocales = locales;
export const BuiltInId = en.id;
export const BuiltInDefault = locales.find((item) => item.id === BuiltInId);
