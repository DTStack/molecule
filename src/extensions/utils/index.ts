import type { IMenuItemProps, IterableItem } from 'mo/types';

export function createMenuDuplicate(item: IMenuItemProps): IMenuItemProps {
    return {
        ...item,
        id: `${item.id}_duplicate`,
        name: `隐藏 “${item.name}”`,
        icon: undefined,
        symbolic: item.id,
    };
}

export function createContextMenu(items: IterableItem[]): IMenuItemProps[] {
    return items.map((item) => ({
        id: item.id,
        name: item.name,
        disabled: item.disabled,
        icon: item.hidden ? undefined : 'check',
    }));
}

export function isInputEle() {
    return document.activeElement instanceof HTMLInputElement || document.activeElement instanceof HTMLTextAreaElement;
}

/**
 * Judge current active element is monaco-editor's textarea or native element
 */
export function isNativeWidget() {
    const isEditableEle =
        document.activeElement instanceof HTMLInputElement || document.activeElement instanceof HTMLTextAreaElement;
    if (!isEditableEle) return true;
    return !document.activeElement?.classList.contains('inputarea');
}
