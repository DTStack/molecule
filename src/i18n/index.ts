export { LocaleService } from './localeService';
export type { ILocaleService } from './localeService';
export type { ILocale } from './localization';
export type { ILocalizeProps } from './localize';
export { localize } from './localize';
