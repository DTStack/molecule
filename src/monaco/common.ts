import { localize } from 'monaco-editor/esm/vs/nls';

export interface IDisposable {
    dispose(): void;
}

export enum KeybindingWeight {
    EditorCore = 0,
    EditorContrib = 100,
    WorkbenchContrib = 200,
    BuiltinExtension = 300,
    ExternalExtension = 400,
}

export const CATEGORIES = {
    View: { value: localize('view', 'View'), original: 'View' },
    Help: { value: localize('help', 'Help'), original: 'Help' },
    Preferences: {
        value: localize('preferences', 'Preferences'),
        original: 'Preferences',
    },
    Developer: {
        value: localize(
            {
                key: 'developer',
                comment: [
                    'A developer on Code itself or someone diagnosing issues in Code',
                ],
            },
            'Developer'
        ),
        original: 'Developer',
    },
};
