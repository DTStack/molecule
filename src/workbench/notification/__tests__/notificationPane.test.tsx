import 'reflect-metadata';
import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import {
    defaultNotificationClassName,
    NotificationPane,
} from '../notificationPane';
import { select } from 'mo/common/dom';
import { expectFnCalled } from '@test/utils';
import { container } from 'tsyringe';
import { ExtendsLocales } from 'mo/extensions/locales-defaults';
import { LocaleService } from 'mo/i18n';

describe('Test NotificationPane Component', () => {
    // initial locales
    const localeService = container.resolve(LocaleService);
    localeService.initialize(ExtendsLocales.contributes!.languages!, 'en');

    test('Match The NotificationPane snapshot', () => {
        const component = renderer.create(
            <NotificationPane data={[{ id: 1, value: '' }]} id="test" />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Render the data', () => {
        const { getByTestId, rerender, queryByText } = render(
            <NotificationPane id="test" />
        );

        expect(select('.mo-notification__body')?.innerHTML).toBe('');

        rerender(
            <NotificationPane
                data={[{ id: 1, value: 'testRender' }]}
                id="test"
            />
        );

        expect(queryByText('testRender')).toBeInTheDocument();

        rerender(
            <NotificationPane
                data={[
                    {
                        id: 1,
                        value: '',
                        render: () => <span data-testid="tid"></span>,
                    },
                ]}
                id="test"
            />
        );
        expect(getByTestId('tid')).toBeInTheDocument();
    });

    test('Display the Notification Pane', () => {
        const { rerender } = render(<NotificationPane id="test" />);

        expect(
            select<HTMLDivElement>('.' + defaultNotificationClassName)!.style
                .display
        ).toBe('none');

        rerender(<NotificationPane id="test" showNotifications />);
        expect(
            select<HTMLDivElement>('.' + defaultNotificationClassName)!.style
                .display
        ).toBe('block');
    });

    test('Listen to the onActionBarClick event', () => {
        expectFnCalled((fn) => {
            const { getByTestId } = render(
                <NotificationPane
                    id="test"
                    onActionBarClick={fn}
                    actionBar={[
                        {
                            id: 'test',
                            'data-testid': 'test',
                        },
                    ]}
                />
            );
            fireEvent.click(getByTestId('test'));
        });
    });

    test('Listen to the onCloseNotification event', () => {
        expectFnCalled((fn) => {
            render(
                <NotificationPane
                    id="test"
                    onCloseNotification={fn}
                    data={[
                        {
                            id: 1,
                            value: '',
                            render: () => <span data-testid="tid"></span>,
                        },
                    ]}
                />
            );
            const testDom = select<HTMLDivElement>('.mo-notification--close');
            fireEvent.click(testDom!);
        });
    });
});
