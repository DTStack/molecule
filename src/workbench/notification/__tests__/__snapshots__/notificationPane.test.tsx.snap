// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NotificationPane Component Match The NotificationPane snapshot 1`] = `
<div
  className="mo-notification mo-context-view--shadow"
  style={
    Object {
      "display": "none",
    }
  }
>
  <header
    className="mo-notification__header"
  >
    <span>
      Notifications
    </span>
    <div
      className="mo-action-bar"
    >
      <ul
        className="mo-action-bar__container"
      />
    </div>
  </header>
  <div
    className="mo-notification__body"
  >
    <div
      className="mo-notification__item"
    >
      <span
        className="mo-notification--close codicon codicon-close"
        onClick={[Function]}
        title="Clear Notification"
      />
    </div>
  </div>
</div>
`;
